### PR TITLE
Guarantee preview/previewClear ordering

### PR DESCRIPTION
**Pull request by neeshy**
This is a followup to the discussion in https://github.com/gokcehan/lf/pull/531 (see: https://github.com/gokcehan/lf/pull/531#issuecomment-751147416)

I have successfully been able to produce a good ordering on calls to the preview and cleaner scripts. No matter how much delay I add to either script, the order always looks like this in my testing:
```
begin pv <first_path> <w> <h> <x> <y>
end pv <first_path> <w> <h> <x> <y>
begin cls <first_path>
end cls <first_path>
begin pv <second_path> <w> <h> <x> <y>
end pv <second_path> <w> <h> <x> <y>
...
```

This solution also only blocks threads responsible for waiting on the previewer script and executing the cleaner script, so these changes don't block the UI itself.

I've also included a commit which passes the file path to the cleaner script.

### CODE REVIEW

**Comment by neeshy** [comment]
Actually, it's a bit racey, and on occasion it will clear before previewing. I'm marking this a draft for now.

**Comment by neeshy** [comment]
Alright, I rewrote this PR such that instead of using two syncing channels, I factored the preview handling into a separate loop which reads from a channel. Preview handling and clearing are effectively handling in one thread now.

File: nav.go [hunk 1]
```diff
@@ -457,7 +459,28 @@ func (nav *nav) position() {
...
+	for {
+		p, ok := <-nav.previewChan
+		if !ok {
+			return
+		}
+		if len(p) != 0 {
+			win := ui.wins[len(ui.wins)-1]
+			nav.previewGen(p, win)
+			path = p
+		} else if len(gOpts.cleaner) != 0 && len(gOpts.cleaner) != 0 && nav.volatilePreview {
```
Comment (gokcehan) [inline nav.go:473]: Is is the same conditional twice?

File: nav.go [hunk 2]
```diff
@@ -457,7 +459,28 @@ func (nav *nav) position() {
...
+			cmd := exec.Command(gOpts.cleaner, path)
```
Comment (gokcehan) [inline nav.go:474]: Isn't `path` empty here?

File: nav.go [hunk 3]
```diff
@@ -525,53 +548,46 @@ func (nav *nav) preview(path string, win *win) {
 	}
 }
 
-func (nav *nav) previewClear() {
-	if len(gOpts.cleaner) != 0 && nav.volatilePreview {
-		nav.volatilePreview = false
+func (nav *nav) preview(path string) {
```
Comment (gokcehan) [inline nav.go:551]: Maybe get rid of single line function here?

File: nav.go [hunk 4]
```diff
@@ -525,53 +548,46 @@ func (nav *nav) preview(path string, win *win) {
...
+	nav.previewChan <- path
+}
 
-		cmd := exec.Command(gOpts.cleaner)
-		if err := cmd.Run(); err != nil {
-			log.Printf("cleaning preview: %s", err)
-		}
-	}
+func (nav *nav) previewClear() {
```
Comment (gokcehan) [inline nav.go:555]: Same as above, putting it inline can make it easier to follow I think.

File: nav.go [hunk 5]
```diff
@@ -457,7 +459,28 @@ func (nav *nav) position() {
...
+			cmd := exec.Command(gOpts.cleaner, path)
+			if err := cmd.Run(); err != nil {
+				log.Printf("cleaning preview: %s", err)
+			}
+			nav.volatilePreview = false
+		}
+	}
+}
+
+func (nav *nav) previewGen(path string, win *win) {
```
Comment (gokcehan) [inline nav.go:483]: If you decide to get rid of the single line functions, maybe you can rename this back to `preview`.

File: app.go [hunk 6]
```diff
@@ -25,7 +25,7 @@ type app struct {
 	ui            *ui
 	nav           *nav
 	ticker        *time.Ticker
-	quitChan      chan bool
+	quitChan      chan struct{}
```
Comment (gokcehan) [inline app.go:28]: What is the purpose of this change?

**Review by gokcehan — COMMENTED** [review]
@neeshy I have tried the patch and I did not experience any issues. I'm not testing this with image previews myself as I'm not very interested in it. We can merge this and let other users report issues. I have marked a few minor points in the patch. I think we can merge it when you are ready.

File: nav.go (hunk 1 above)
Comment (neeshy) [reply nav.go]: Whoops, I meant to put gOpts.previewer for the first one. Fixed.

File: nav.go (hunk 2 above)
Comment (neeshy) [reply nav.go:474]: It shouldn't be, since nav.volatilePreview is part of the condition of the second branch, and that variable only gets set during calls to previewGen (which is is only called within previewLoop). So, the second branch of that if statement will only be evaluated after iterations where previewGen is called and the path is set.

File: nav.go (hunk 3 above)
Comment (neeshy) [reply nav.go]: OK, done.

File: nav.go (hunk 4 above)
Comment (neeshy) [reply nav.go]: OK.

File: nav.go (hunk 5 above)
Comment (neeshy) [reply nav.go]: Right, done.

File: app.go (hunk 6 above)
Comment (neeshy) [reply app.go:28]: I was originally running `preview` and `previewClear` in their own respective threads and synchronizing them through two `chan struct{}` variables. I changed all the instances of `chan bool` to be consistent with this. Using `chan struct{}` is idiomatic Go and it doesn't cause an allocation (since struct{} has a size of 0) whereas a `chan bool` will.

**Comment by gokcehan** [comment]
@neeshy It seems good, thank you.
